cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -492,25 +491,26 @@ pub(crate) fn lower_address(
 
     // If there is any offset, load that first into `addr`, and add the `reg`
     // that we kicked out of the `AMode`; otherwise, start with that reg.
-    if offset != 0 {
-        lower_add_immediate(ctx, addr, reg, offset)
+    let init = if offset != 0 {
+        lower_add_immediate(ctx, reg, offset)
     } else {
-        ctx.emit(Inst::gen_move(addr, reg, I64));
-    }
+        reg
+    };
 
     // Now handle reg64 and reg32-extended components.
-    lower_add_addends(ctx, addr, addends64, addends32);
+    let init = lower_add_addends(ctx, init, addends64, addends32);
+    ctx.emit(Inst::gen_move(addr, init, I64));
```

Comment:
The move is slightly awkward here; is there a way we could rewrite the `AMode` we created above to use `init` directly? (I know this is kind of awkward because we need to do the above match first to get the `reg` addend, but perhaps we could have a `.replace_first_reg(..)` on the `AMode` or something like that.)